### PR TITLE
refactor build script

### PR DESCRIPTION
**Pull request by frg2089**
Link CnCNet/xna-cncnet-client#341

### CODE REVIEW

**Comment by Rans4ckeR** [comment]
I get this error:

> .\Build-All.ps1
Building Ares UniversalGL Release net6.0  +  any...
InvalidOperation: Cannot index into a null array.
InvalidOperation: Cannot index into a null array.
Write-Debug: A positional parameter cannot be found that accepts argument '+'.
Write-Debug: A positional parameter cannot be found that accepts argument '+'.

**Comment by frg2089** [comment]
> I get this error:

I fixed this errors. [CI status is here](https://github.com/frg2089/xna-cncnet-client/actions/runs/3064776361/jobs/4948212155)

**Comment by frg2089** [comment]
@Metadorius I've refactored the build script here, and you can make some suggestions and requests here.

**Comment by Rans4ckeR** [comment]
Can you update the list of output files like this?

remove from \net6.0-windows10.0.22000.0\Resources\Binaries\Windows
ClientUpdater.dll
DiscordRPC.dll
DiscordRPC.pdb
DiscordRPC.xml
Localization.dll
Localization.pdb
lzo.net.dll
Microsoft.CodeAnalysis.CSharp.dll
Microsoft.CodeAnalysis.dll
Microsoft.CodeAnalysis.VisualBasic.dll
Microsoft.Extensions.DependencyInjection.Abstractions.dll
Microsoft.Windows.SDK.NET.dll
Newtonsoft.Json.dll
OpenMcdf.dll
Rampastring.Tools.dll
SixLabors.ImageSharp.dll
System.Management.dll
TextCopy.dll
WinRT.Runtime.dll

add to \net48\Resources
clientdx.exe.config
clientdx.pdb
clientogl.exe.config
clientogl.pdb
clientxna.exe.config
clientxna.pdb

remove from \net48\Resources\Binaries
clientogl.exe.config
clientxna.exe.config

remove from \net48\Resources\Binaries\OpenGL
clientogl.pdb
libopenal.1.dylib
libSDL2-2.0.0.dylib

remove from \net48\Resources\Binaries\OpenGL\x64
libopenal.so.1
libSDL2-2.0.so.0

**Comment by frg2089** [comment]
I found that compiling in a Linux environment generates a clientogl executable, have you dealt with this file?

**Comment by Rans4ckeR** [comment]
> I found that compiling in a Linux environment generates a clientogl executable, have you dealt with this file?

Yes:
https://github.com/Rans4ckeR/xna-cncnet-client/blob/6ff904f8560f063c1e3ffd0114a68d894b316398/BuildScripts/lib/BuildCore.ps1#L93-L95

On linux we have
`\any\Resources\Binaries\OpenGL\clientogl.dll`
and
AppHostLauncher: `\any\Resources\Binaries\clientogl.`

on Windows we have
`\any\Resources\Binaries\OpenGL\clientogl.dll`
and
AppHostLauncher: `\any\Resources\Binaries\clientogl.exe`

The current script handles it this way I believe:
Delete the AppHostLauncher file when building for `any` because the executable is not cross-platform.
Keep the AppHostLauncher file when building for a specific platform, move it to \Resources.

When building for a specific platform e.g. `-r linux-x64` the output with the current scripts appears in `\net6.0\linux-x64` instead of `\net6.0\any`. `any` represents the cross-platform binaries folder.

@frg2089 updated above reply

**Comment by Rans4ckeR** [comment]
Can you still remove these files from the build output? They don't do any harm but do take up several MBs.

\net48\Resources\Binaries\OpenGL
libopenal.1.dylib
libSDL2-2.0.0.dylib

\net48\Resources\Binaries\OpenGL\x64
libopenal.so.1
libSDL2-2.0.so.0

\net6.0-windows10.0.22000.0\Resources\Binaries\Windows
Microsoft.CodeAnalysis.CSharp.dll
Microsoft.CodeAnalysis.dll
Microsoft.CodeAnalysis.VisualBasic.dll

**Comment by frg2089** [comment]
I used a Task from the .Net SDK to implement the AppHost Hack at commit 313aeb3.

What do you think of this solution?

**Comment by Rans4ckeR** [comment]
> I used a Task from the .Net SDK to implement the AppHost Hack at commit [313aeb3](https://github.com/Rans4ckeR/xna-cncnet-client/commit/313aeb38180740348da3619623d6c8047b6f8094).
> 
> What do you think of this solution?

It would be a better solution than the previous one if it worked on all platforms.
It seems to work fine when building for Windows, but it fails for linux:

>The application to execute does not exist: /net6.0/any/Resources/Binaries\OpenGL/clientogl.dll

This slash is wrong: `Binaries\OpenGL`

**Comment by frg2089** [comment]
> This slash is wrong: `Binaries\OpenGL`

I fixed it at e85cf81.

**Comment by frg2089** [comment]
> Can you still remove these files from the build output? They don't do any harm but do take up several MBs.
> 
> \net48\Resources\Binaries\OpenGL libopenal.1.dylib libSDL2-2.0.0.dylib
> 
> \net48\Resources\Binaries\OpenGL\x64 libopenal.so.1 libSDL2-2.0.so.0
> 
> \net6.0-windows10.0.22000.0\Resources\Binaries\Windows Microsoft.CodeAnalysis.CSharp.dll Microsoft.CodeAnalysis.dll Microsoft.CodeAnalysis.VisualBasic.dll

i finished this issue.

**Comment by Metadorius** [comment]
image
I haven't reviewed the full code, but you have to understand, I suggested that scripts should be simplified, and from a brief look you are going in the opposite direction. For example, what benefit does using PS modules have here? There was no problem that this solves.

Everything became much more verbose and complicated for a simple task of just building the correct configuration and copying the files to correct places.

**Comment by frg2089** [comment]
> image I haven't reviewed the full code, but you have to understand, I suggested that scripts should be simplified, and from a brief look you are going in the opposite direction. For example, what benefit does using PS modules have here? There was no problem that this solves.
> 
> Everything became much more verbose and complicated for a simple task of just building the correct configuration and copying the files to correct places.

This script is simple enough now.

The script is now built directly from the entry script calling the generic build function within the module.

Entry Script -> Generic Build Function -> Build Function

**Comment by Rans4ckeR** [comment]
> image I haven't reviewed the full code, but you have to understand, I suggested that scripts should be simplified, and from a brief look you are going in the opposite direction. For example, what benefit does using PS modules have here? There was no problem that this solves.
> 
> Everything became much more verbose and complicated for a simple task of just building the correct configuration and copying the files to correct places.

How about an approach where we do all processing using msbuild instead of PS such as below?
- #14

**Comment by frg2089** [comment]
> How about an approach where we do all processing using msbuild instead of PS such as below?

It is right to put complex judgments into the MSBuild project that
The build script should be kept simple and easy to modify.
So I abstracted a build function from other scripts to make it easier to modify later.
I thought I should use long parameter names in the script, so I replaced all short parameter names.

File: BuildScripts/Common.ps1 [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+$CompiledRoot = Join-Path $RepoRoot Compiled
+$EngineMap = @{
+  'UniversalGL' = 'OpenGL'
+  'WindowsDX'   = 'Windows'
+  'WindowsGL'   = 'OpenGL'
+  'WindowsXNA'  = 'XNA'
+}
+
+function Build-Project($Configuration, $Game, $Engine, $Framework) {
+  $Output = Join-Path $CompiledRoot $Game $Framework Resources Binaries ($EngineMap[$Engine])
```
Comment (Rans4ckeR) [inline BuildScripts/Common.ps1]: Can you add `any` for UniversalGL?

`$path\..\Compiled\YR\net6.0\any\Resources\Binaries\OpenGL`

**Review by Rans4ckeR — CHANGES_REQUESTED** [review]

**Review by Rans4ckeR — APPROVED** [review]

**Pull request merged**